### PR TITLE
💊 Add Is Valid Url Constraint

### PR DESCRIPTION
**Pull request by nerg4l**
This PR adds a `Constraint` that can be used in tests to see if the given value is a valid URL.

Can be used with `assertThat`.

### CODE REVIEW

**Comment by yak0d3** [comment]
One small recommendation, why not creating a method for the constraint, for instance `assertValidUrl`?

**Comment by axlon** [comment]
@nerg4l I added some extra stuff:
- Moved assertRedirectsTo logic to its own constraint class
- Renamed FollowsRedirects trait to CustomAssertions with the assertValidUrl and assertRedirectsTo functions
- Removed assertInternalType calls, because when we assert if something is a valid url or if it redirects to a certain destination, we already know it's a string (and it's also triggering deprecation warnings in PHPUnit 8)

**Comment by nerg4l** [comment]
@axlon Instead of `CustomAssertions` it would be better to name the trait `UrlAssertions`. Also we should discuss these changes first because I was already doing the same.

File: tests/Constraints/RedirectsTo.php [hunk 1]
```diff
@@ -19,7 +19,7 @@ class RedirectsTo extends Constraint
      * @param int $redirects
      * @return void
      */
-    public function __construct($destination, int $redirects)
+    public function __construct($destination, int $redirects = 1)
```
Comment (axlon) [inline tests/Constraints/RedirectsTo.php]: This serves no purpose, the constructor is never called directly

File: tests/Concerns/UrlAssertions.php [hunk 2]
```diff
@@ -27,6 +27,16 @@ public static function assertRedirectsTo($expected, $actual, int $redirects = 1)
      */
     public static function assertValidUrl($actual)
     {
-        static::assertThat($actual, new IsValidUrl);
+        static::assertThat($actual, static::isValidUrl());
+    }
+
+    public static function isValidUrl(): IsValidUrl
```
Comment (axlon) [inline tests/Concerns/UrlAssertions.php]: Add doc blocks here

File: tests/Concerns/UrlAssertions.php [hunk 3]
```diff
@@ -27,6 +27,16 @@ public static function assertRedirectsTo($expected, $actual, int $redirects = 1)
...
+    {
+        return new IsValidUrl;
+    }
+
+    public static function redirectsTo($expected, int $redirects = 1): RedirectsTo
```
Comment (axlon) [inline tests/Concerns/UrlAssertions.php]: Add doc blocks here

**Comment by axlon** [comment]
@nerg4l my bad, I wasn't aware. I have some comments on the new changes (only minor things)

**Comment by yak0d3** [comment]
As for the naming of the trait, IMO something like `HasUrlAssertions` would look better for a 'concern'.

**Comment by axlon** [comment]
I agree, traits usually are named after the action/feature they provide, so HasUrlAssertions or AssertsUrls would probably be better

File: tests/Constraints/RedirectsTo.php (hunk 1 above)
Comment (nerg4l) [reply tests/Constraints/RedirectsTo.php]: I added this for consistency. If the assert method has this as a default value, then the constructor also should have this as the default value.

Maybe this one comes to believes so if it bothers you I can revert it.

**Comment by nerg4l** [comment]
`HasUrlAssertions` is the name then.

**Pull request merged**